src/main/java/net/imagej/ops/DefaultOpMatchingService.java

```diff
@@ -246,6 +244,220 @@ public Module assignInputs(final Module module, final Object... args) {
 	}
 
 	/**
+	 * Gets a list of valid candidates injected with padded arguments.
+	 * <p>
+	 * Helper method of {@link #filterMatches}.
+	 * </p>
+	 * 
+	 * @param candidates list of candidates
+	 * @return a list of valid candidates with arguments injected
+	 */
+	private List<OpCandidate<?>> validCandidates(
+		final List<OpCandidate<?>> candidates)
+	{
+		final ArrayList<OpCandidate<?>> validCandidates = new ArrayList<>();
+		for (final OpCandidate<?> candidate : candidates) {
+			if (!valid(candidate) || !outputsMatch(candidate)) continue;
+			final Object[] args = padArgs(candidate);
+			if (args == null) continue;
+			candidate.setArgs(args);
+			if (missArgs(candidate)) continue;
+			validCandidates.add(candidate);
+		}
+		return validCandidates;
+	}
+
+	/**
+	 * Extracts a list of candidates that requires lossless conversion to match
+	 * with the reference.
+	 * <p>
+	 * Helper method of {@link #filterMatches(List)}.
+	 * </p>
+	 */
+	@SuppressWarnings("unused")
+	private List<OpCandidate<?>> losslessMatches(
+		final List<OpCandidate<?>> candidates)
+	{
+		final ArrayList<OpCandidate<?>> matches = new ArrayList<>();
+		double priority = Double.NaN;
+		for (final OpCandidate<?> candidate : candidates) {
+			final ModuleInfo info = candidate.cInfo();
+			final double p = info.getPriority();
+			if (p != priority && !matches.isEmpty()) {
+				// NB: Lower priority was reached; stop looking for any more matches.
+				break;
+			}
+			priority = p;
+
+			if (!losslessMatch(candidate)) continue;
+
+			if (moduleConforms(candidate)) matches.add(candidate);
+		}
+		return matches;
+	}
+
+	/**
+	 * Determines if the candidate arguments match with lossless conversion. Needs
+	 * support from the conversion in the future.
+	 */
+	@SuppressWarnings("unused")
+	private boolean losslessMatch(final OpCandidate<?> candidate) {
+		// NB: Not yet implemented
+		return false;
+	}
+
+	/**
+	 * Extracts a list of candidates that requires lossy conversion to match with
+	 * the reference.
+	 * <p>
+	 * Helper method of {@link #filterMatches(List)}.
+	 * </p>
+	 */
+	private List<OpCandidate<?>> lossyMatches(
+		final List<OpCandidate<?>> candidates)
+	{
+		final ArrayList<OpCandidate<?>> matches = new ArrayList<>();
+		double priority = Double.NaN;
+		for (final OpCandidate<?> candidate : candidates) {
+			final ModuleInfo info = candidate.cInfo();
+			final double p = info.getPriority();
+			if (p != priority && !matches.isEmpty()) {
+				// NB: Lower priority was reached; stop looking for any more matches.
+				break;
+			}
+			priority = p;
+
+			if (typesMatch(candidate, candidate.getArgs()) != -1) continue;
```

Comment:
This method and `losslessMatches` and `perfectMatches` are all the same, except for this type matching test per candidate. Can we use some Java 8 functional goodness here to consolidate them all into a single method?